src/Polly.Core/CircuitBreaker/Health/RollingHealthMetrics.cs

```diff
@@ -0,0 +1,72 @@
+namespace Polly.CircuitBreaker.Health;
+
+internal sealed class RollingHealthMetrics : HealthMetrics
+{
+    private readonly TimeSpan _samplingDuration;
+    private readonly TimeSpan _windowDuration;
+    private readonly Queue<HealthWindow> _windows;
+
+    private HealthWindow? _currentWindow;
+
+    public RollingHealthMetrics(TimeSpan samplingDuration, short numberOfWindows, TimeProvider timeProvider)
+        : base(timeProvider)
+    {
+        _samplingDuration = samplingDuration;
+        _windowDuration = TimeSpan.FromTicks(_samplingDuration.Ticks / numberOfWindows);
+        _windows = new Queue<HealthWindow>();
+    }
+
+    public override void IncrementSuccess() => UpdateCurrentWindow().Successes++;
+
+    public override void IncrementFailure() => UpdateCurrentWindow().Failures++;
+
+    public override void Reset()
+    {
+        _currentWindow = null;
+        _windows.Clear();
+    }
+
+    public override HealthInfo GetHealthInfo()
+    {
+        UpdateCurrentWindow();
+
+        var successes = 0;
+        var failures = 0;
+        foreach (var window in _windows)
+        {
+            successes += window.Successes;
+            failures += window.Failures;
```

Comment:
Should we worry about being atomic with incrementing these counters?

Reply:
All these operations are executed under a lock so atomicity is ensured.